llama_stack/providers/inline/post_training/huggingface/recipes/finetune_single_device.py

```diff
@@ -0,0 +1,435 @@
+            num_train_epochs=config.n_epochs,
+            per_device_train_batch_size=config.data_config.batch_size,
+            fp16=device.type == "cuda",
+            bf16=device.type != "cuda",
+            use_cpu=False,  # TODO: investigate, this causes a segfault. device.type == "cpu",
+            save_strategy=save_strategy,
+            save_steps=save_steps,
+            report_to="none",
+            max_seq_length=provider_config.max_seq_length,
+            gradient_accumulation_steps=config.gradient_accumulation_steps,
+            gradient_checkpointing=provider_config.gradient_checkpointing,
+            learning_rate=lr,
+            warmup_ratio=provider_config.warmup_ratio,
+            weight_decay=provider_config.weight_decay,
+            logging_steps=logging_steps,
+            # Enable validation
+            eval_strategy="steps",
+            eval_steps=eval_steps,
+            save_total_limit=provider_config.save_total_limit,
+            remove_unused_columns=False,
+            dataloader_pin_memory=provider_config.dataloader_pin_memory,
+            dataloader_num_workers=provider_config.dataloader_num_workers,
+            dataset_text_field="text",
+            packing=False,
+            # Add evaluation metrics
+            load_best_model_at_end=True if output_dir_path else False,
+            metric_for_best_model="eval_loss",
+            greater_is_better=False,
+        )
+
+        # Initialize trainer with both train and eval datasets
+        trainer = SFTTrainer(
+            model=model_obj,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            peft_config=peft_config,
+            args=training_arguments,
+        )
+
+        # Train
+        logger.info("Starting training")
+        try:
+            trainer.train()
+            memory_stats["after_training"] = get_memory_stats(device)
+
+            # Save final model
+            model_obj.config.use_cache = True
+            if lora_config:
+                model_obj = trainer.model.merge_and_unload()
+            else:
+                model_obj = trainer.model
+
+            checkpoint = None
+            checkpoints = None
+            if output_dir_path:
+                model_obj.save_pretrained(output_dir_path / "merged_model")
+
+                # Create checkpoint
+                checkpoint = Checkpoint(
+                    identifier=f"{model}-sft-{config.n_epochs}",
+                    created_at=datetime.now(timezone.utc),
+                    epoch=config.n_epochs,
+                    post_training_job_id=job_uuid,
+                    path=str(output_dir_path / "merged_model"),
+                )
+                checkpoints = [checkpoint]
+
+            return memory_stats, checkpoints
+        finally:
+            # Clean up resources
+            if hasattr(trainer, "model"):
+                if device.type != "cpu":
```

Comment:
lines 408-414 are implemented in torchtune. Please move the code to a common function (`evacuate_model_from_device`?)

Reply:
created common/utils.py for post_training and `clear_model`

Reply:
I am lazy importing torch so we don't hit issues within this method.

Reply:
I like the name @booxter suggested more than `clear_model`

Reply:
sure I can change it

Reply:
changed